internal/cmd/gocdk/fswatch.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// waitForDirChange blocks until the directory changes or the context's Done
+// channel is closed.
+func waitForDirChange(ctx context.Context, path string, pollInterval time.Duration) error {
+	tick := time.NewTicker(pollInterval)
+	defer tick.Stop()
+	init, err := snapshotTree(path)
+	if err != nil {
+		return xerrors.Errorf("wait for %s to change: %w", path, err)
+	}
+	for {
+		select {
+		case <-tick.C:
+			same, err := init.sameAs(path)
+			if err != nil {
+				// TODO(light): Log error.
+				continue
+			}
+			if !same {
+				return nil
+			}
+		case <-ctx.Done():
+			return xerrors.Errorf("wait for %s to change: %w", path, ctx.Err())
+		}
+	}
+}
+
+// dirSnapshot is a snapshot of metadata for a directory.
+type dirSnapshot struct {
+	entries []*dirSnapshotEntry // sorted by name
+}
+
+// newDirSnapshot converts a list of os.FileInfo into a directory snapshot.
+// It does not fill in any of the entries' snapshot fields.
+func newDirSnapshot(contents []os.FileInfo) *dirSnapshot {
+	snap := &dirSnapshot{
+		entries: make([]*dirSnapshotEntry, 0, len(contents)),
+	}
+	for _, info := range contents {
+		name := info.Name()
+		if isIgnoredForWatch(name) {
+			continue
+		}
+		snap.entries = append(snap.entries, &dirSnapshotEntry{
+			name:    name,
+			mode:    info.Mode(),
+			modTime: info.ModTime(),
+		})
+	}
+	sort.Slice(snap.entries, func(i, j int) bool {
+		return snap.entries[i].name < snap.entries[j].name
+	})
+	return snap
+}
+
+// snapshotTree recursively snapshots a directory.
+func snapshotTree(path string) (*dirSnapshot, error) {
+	// Unfortunately filepath.Walk does not work well for this because
+	// we need to keep track of the parent snapshot variable to write to.
```

Comment:
I did not understand this comment, can you clarify? Aren't we always appending to the same `dirSnapshot.entries`?

Reply:
I see, no. Could it be done that way? I.e., flatten the recursive directory structure into `entries` (where the entry name is the full path from the root). Then the equality comparisons are much easier.

Reply:
I didn't do it that way because I was worried about path normalization issues, but I think the two approaches are roughly the same. Do you think that gives a significant benefit for readability?

Reply:
Done.